[Config] Use better typehint in PHP Configuration

jderusse opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

This PR now followup #46328 to replace the generic `mixed` type-hint by type inferred from normalizations (ie, if a config use `ifString`, when now that the normalization works only for string, these `mixed` can safely be replaced by `string`)

jderusse committed:
Allow scalar configuration in PHP Configuration

ro0NL commented:
Related to #43164? 

> Draw back is, the signature of the method place is now place(mixed $value = []): self|\Symfony\Config\Framework\Workflows\WorkflowsConfig\PlaceConfig Is it an issue?

IMHO it's a step back yes.

I shortcuts worth pursuing? Rather than fully qualified/explicit config?

> But sadly lack the previous behavior. People have to enter the full configuration.

works for me :)

jderusse commented:
Actually, I converted a personal application to PHP configuration, and have to admit that I've been disappointed by the experience.

Some really simple YAML configuration `framework.lock: flock` become obscure `$framework->lock()->resource('default', ['flock'])` <= why `resource`, why `default` why `flock` is wrapped into an array?

Some configurations short `monolog.handler.main.excluded_http_codes: [403, 404]` become super verbose with intermediate varianbles
```php
    $mainHandler = $monolog->handler('main');
    $mainHandler->excludedHttpCode()->code(403);
    $mainHandler->excludedHttpCode()->code(404);

```

Moreover, when I switched my configuration, the symfony documentation was not-yet updated, and I had to dig into the `Configuration.php` file to understand how to convert my yaml file into PHP... Now this will be easier for symfony (because most of config are not documented) but this is not the case for many other bundles.

If the issue is only about the double return signature `xxx(): self|child` we could keep the current behavior, and throw an exception when people try to update a property inside the shortcut value
```
$framework->lock('redis://localhost')->resource(...); <== InvalidConfigurationException
```

nicolas-grekas commented:
Rebase needed, in case there is anything left after  #46328

nicolas-grekas commented:
Friendly ping @jderusse 
/cc @HypeMC

Is this PR still needed after https://github.com/symfony/symfony/pull/46328?

HypeMC commented:
@nicolas-grekas This PR provides some additional features that #46328 doesn't have, eg allowed type hints on generated methods:
https://github.com/symfony/symfony/blob/962834b63066c1be04bed19cbc6762d3136bcfa3/src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes/Symfony/Config/ScalarNormalizedTypesConfig.php#L25

Currently `mixed` is used in these situations:
https://github.com/symfony/symfony/blob/5bf31c57c3d07ab6a04bdeb8fdc2957e0591dcb5/src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes/Symfony/Config/ScalarNormalizedTypesConfig.php#L31

My PR only covers using scalars with the `beforeNormalization` hook.

nicolas-grekas commented:
Would you be able to take over the PR @HypeMC? I feel like @jderusse would appreciate :)

jderusse commented:
This PR also adds an (opiniated) change in the returned object:
When people configure an entry with a scalar value, then, the returned value is the object and not the child object.
This allows chaining "aliases". ie
```diff
     $subscriptionWorkflow = $framework->workflows()->workflows('subscription');

     $subscriptionWorkflow->place(Subscription::STATUS_SUBMITTED);
-    $subscriptionWorkflow->place(Subscription::STATUS_CANCELED);
-    $subscriptionWorkflow->place(Subscription::STATUS_ACCEPTED);
-    $subscriptionWorkflow->place(Subscription::STATUS_VALIDATED);
+      ->place(Subscription::STATUS_CANCELED)
+      ->place(Subscription::STATUS_ACCEPTED)
+      ->place(Subscription::STATUS_VALIDATED);
```

HypeMC commented:
> Would you be able to take over the PR @HypeMC? I feel like @jderusse would appreciate :)

Sure, if @jderusse doesn't mind, I'll take over.

jderusse commented:
> if @jderusse doesn't mind,

no issue at all, thanks for your help

jderusse commented:
PR rebased and duplicated logic with #46328 removed

chalasr reviewed (changes requested):
Missing CHANGELOG entry. 👍 to ship this in 6.2

nicolas-grekas commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -212,6 +213,24 @@ public function setNormalizationClosures(array $closures)
     }
 
+    /**
+     * Sets the list of type supported by normalization.
```
s/type/types

nicolas-grekas commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -212,6 +213,24 @@ public function setNormalizationClosures(array $closures)
 
+    /**
+     * Sets the list of type supported by normalization.
+     * see ExprBuilder::TYPE_* constants.
```
let's add a blank line before this one

nicolas-grekas commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -212,6 +213,24 @@ public function setNormalizationClosures(array $closures)
+    }
+
+    /**
+     * Gets the list of type supported by normalization.
```
types

nicolas-grekas commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -212,6 +213,24 @@ public function setNormalizationClosures(array $closures)
+
+    /**
+     * Gets the list of type supported by normalization.
+     * see ExprBuilder::TYPE_* constants.
```
blank line

nicolas-grekas commented on src/Symfony/Component/Config/Definition/Builder/ExprBuilder.php:
```diff
@@ -21,7 +21,13 @@
+    public const TYPE_ARRAY = 'array';
+
     protected $node;
+    public $allowedTypes;
```
blank line before

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ConfigBuilderGenerator.php:
```diff
@@ -209,19 +216,21 @@ private function handlePrototypedArrayNode(PrototypedArrayNode $node, ClassBuild
-        if (null !== $parameterType || $prototype instanceof ScalarNode) {
+        $nodeParameterTypes = $this->getParameterTypes($node);
+        $prototypeParameterTypes = $this->getParameterTypes($prototype);
+        if (!($prototype instanceof ArrayNode && (!$prototype instanceof PrototypedArrayNode || !$prototype->getPrototype() instanceof ScalarNode))) {
```
`if (!$prototype instanceof ArrayNode || ($prototype instanceof PrototypedArrayNode && $prototype->getPrototype() instanceof ScalarNode)) {`

nicolas-grekas commented:
Thank you @jderusse.

alexndlm commented:
@jderusse

Before in `\Symfony\Config\Framework\HttpClient\ScopedClientConfig\RetryFailedConfig::httpCode` was:
```php
public function httpCode(string $code, mixed $value = []): \Symfony\Config\Framework\HttpClient\ScopedClientConfig\RetryFailed\HttpCodeConfig|static
```

Now it generates:
```php
public function httpCode(string $code, array $value = []): \Symfony\Config\Framework\HttpClient\ScopedClientConfig\RetryFailed\HttpCodeConfig|static
```

Is this a bug in the config generator or in the HttpClient?
